### DISCUSSION TITLE
.foreach, for (let i = 0; i < length; i++) etc

### DISCUSSION THREAD

**Discussion by yosiasz** [grafana/grafana#64169] [General]
Greetings,

Perusing around grafana code. I am surprised to see heavy use of old school loops. Could this explain some performance issues when data set is big?

Here are some benchmarks with .forEach and for loops vs modern use of .map and .reduce 

https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/

Is there plan to do some overhaul for grafana to use more performant approach to iterating through data?

Thanks

**Comment by gtk-grafana** [comment]
Thanks @yosiasz, this is a good idea, but it's hard to prioritize something like this without any evidence of the impact. Looking around at the code, most of the use of foreach is in unit tests, and other usages are in loops with a constant size, so rewriting in a more performant loop would provide negligible benefit.

If there were specific usages of foreach that you can demonstrate are slowing things down, that would likely help this issue get prioritized. 

Otherwise, the community is more then welcome to submit PRs to fix things like this, but Grafana labs will likely not prioritize this issue until it's apparent that it will benefit our users, and more so then other ongoing/planned work.
